### PR TITLE
Add overload for setState to support function as argument

### PR DESCRIPTION
**Pull request by Alxswan**
Hi! I'm a junior developer and this is my first open source PR so please be gentle.

I have updated setState so that it can also take a function as per the React.js docs: https://facebook.github.io/react/docs/state-and-lifecycle.html

I'm not sure if I should add another test scenario to catch this change or if updating the test expectations as I have done is enough.

### CODE REVIEW

**Comment by CodyReichert** [comment]
I didn't even know that setState could do that. Nice job!

File: lib/react.js [hunk 1]
```diff
@@ -22,8 +22,8 @@ declare class React$Component<DefaultProps, Props, State> {
 
   // action methods
 
-  // TODO: partialState could be a function as well
   setState(partialState: $Shape<State>, callback?: () => void): void;
+  setState(callback: (prevState: State, props: Props ) => $Shape<State>): void;
```
Comment (nmn) [inline lib/react.js]:
```typescript
setState(callback: (prevState: State, props: Props ) => $Shape<State>, callback?: () => void): void;
```

**Comment by apsavin** [comment]
@Alxswan Are you planning to finish this PR?

@nmn This PR actually fixes a bug. Can you add an appropriate label? 
If @Alxswan will not respond, I think I'll create another PR. 

By the way, why React (and some other libs) type definitions are here, not in https://github.com/flowtype/flow-typed ?

**Comment by Alxswan** [comment]
@nmn @apsavin PR updated :)

**Comment by goodmind** [comment]
This was fixed in https://github.com/facebook/flow/commit/95d6d11a88235a2106649aaa1afc134f0340c9da#diff-5ca8a047db3f6ee8d65a46bba4471236

**Pull request closed without merging**